### PR TITLE
strutils: fix buffer overflow in formatBiggestFloat

### PR DESCRIPTION
**Pull request by alaviss**
## Summary
Fixed a potential buffer overflow in `struts.formatBiggestFloat`.

## Details
Previously,  `strutils.formatBiggestFloat`  was implemented using 
`sprintf`  and as such came with a risk of buffer overflow. This was
observed via tests in disruptek/insideout.

* Replaced `sprintf` with `snprintf`.
* Remove the 2 KiB temporary buffer in favor of directly outputting to
the target buffer.
* Added extra error checking code for errors from `snprintf`.

### CODE REVIEW

File: lib/pure/strutils.nim [hunk 1]
```diff
@@ -2359,26 +2362,64 @@ func formatBiggestFloat*(f: BiggestFloat, format: FloatFormatMode = ffDefault,
...
+      result = newString(L + 1)
+      L = formatTo(result, result.len.csize_t)
     else:
       frmtstr[1] = floatFormatToChar[format]
       frmtstr[2] = '\0'
-      L = c_sprintf(addr buf, addr frmtstr, f)
-    result = newString(L)
-    for i in 0 ..< L:
+      template formatTo(buf: cstring, size: csize_t): cint =
+        # This has the side-effect of invalidating c_errno, but we also do
```
Comment (saem) [inline lib/pure/strutils.nim:2402]: Is this not duplicated, can we not move it to a common scope?

**Review by saem — APPROVED** [review]
Don't suppose there is an easy enough reproduction for a test, or do you think it's too implementation specific?

File: lib/pure/strutils.nim (hunk 1 above)
Comment (alaviss) [reply lib/pure/strutils.nim:2402]: The two calls are different, one has precision and one hasn't.

**Comment by alaviss** [comment]
> Don't suppose there is an easy enough reproduction for a test, or do you think it's too implementation specific?

It's too implementation specific and won't be of any help if we move to a native impl at one point.

File: lib/pure/strutils.nim (hunk 1 above)
Comment (saem) [reply lib/pure/strutils.nim:2402]: Thanks, was on the phone and I missed that while scrolling up and down.

**Comment by saem** [comment]
/merge

(updated the PR body slightly)

**Pull request merged**